### ISSUE TITLE
custom-server-typescript example is not debuggable

### ISSUE THREAD

**Issue by AssisrMatheus** [zeit/next.js#8621]
# Examples bug report
`custom-server-typescript` is not debuggable

## Example name
custom-server-typescript

## Describe the bug
I've managed to debug next with Typescript if not using a custom server, but I need to use custom server so I can use `next-routes` package. 

I followed the `custom-server-typescript` example but cannot debug the server. I've tried *multiple* configurations found on internet, some work only on the server file and not in react components, others work in almost everything but don't find the sourcemap correctly, but none has worked properly until now.

This problem is recurrent and there are lots of reports of people with this same issue on Spectrum(which, btw, is a deserted land). You can go to https://spectrum.chat/next-js?tab=posts and search for `debug typescript` and see a lot of people with this problem yet unresolved.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Clone `custom-server-typescript`
2. Create a `launch.json` file just like mine
3. Run debugging
4. Try to break inside a react component's get initial props or render method

## Expected behavior
Breakpoints should hit within their respective sourcemapped TS files

## System information

- OS: Windows 10
- Version of Next.js: 9.0.5
- Version of VSCode: 1.37.1
- Version of Typescript: 3.6.2

## Additional context

This is my current `launch.json`

```{
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Node Inspector",
      "type": "node",
      "request": "launch",
      "runtimeArgs": ["--nolazy", "-r", "ts-node/register"],
      "args": ["${workspaceRoot}/server/index.ts"],
      "cwd": "${workspaceRoot}",
      "outFiles": ["${workspaceFolder}/dist/src/**/*.js", "${workspaceFolder}/.next/server/**/*.js"],
      "protocol": "inspector",
      "internalConsoleOptions": "openOnSessionStart",
      "sourceMaps": true,
      "autoAttachChildProcesses": true,
      "env": {
        "TS_NODE_PROJECT": "tsconfig.server.json",
        "TS_NODE_FILES": "true"
      }
    }
  ]
}
```

**Comment by craigpg** [comment]
I was able to debug server-side components (running `yarn dev`) with your setup by creating a `next.config.js` with the following:

```
const webpack = require('webpack')

module.exports = {
  webpack(config) {
    config.devtool = 'eval-source-map';
    return config
  }
}
```

In vscode, the breakpoints are (re)enabled once the server compiles the page.

Looks like the build sets `devTool` to `cheap-module-source-map` (see https://github.com/zeit/next.js/blob/canary/packages/next/build/webpack-config.ts#L191), which seems sufficient for client-side debugging.

I confirmed that client-side debugging still works with the above change.

**Comment by patsadow2** [comment]
Great! It's working @craigpg and @AssisrMatheus 

Do you maybe know how to adjust tsconfig.sever.json to make possible read your own modules?

Debugger attached.
Error: Cannot find module 'common/template'

**Comment by petvas** [comment]
I found a working solution:

TL DR:
`package.json` add this line under `scripts`
```js
"debug": "set NODE_OPTIONS=--inspect&&next",
```

`.vscode/launch.json`
```js
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "type": "chrome",
            "request": "launch",
            "name": "Next: Chrome",
            "url": "http://localhost:3000",
            "webRoot": "${workspaceFolder}",
        },
        {
            "type": "node",
            "request": "launch",
            "name": "Next: Node",
            "runtimeExecutable": "npm",
            "runtimeArgs": [
                "run",
                "debug",
            ],
            "port": 9229,
            "console": "integratedTerminal",
        }
    ],
    "compounds": [
        {
            "name": "Next: Full",
            "configurations": ["Next: Node", "Next: Chrome"]
        }
    ]
}
```

`next.config.js`
```js
module.exports = {
    /*...*/
    webpack: (config) => {
        config.output = config.output || {};
        config.output.devtoolModuleFilenameTemplate = function(info){
            return "file:///"+encodeURI(info.absoluteResourcePath);
        }
    },
}
```

The issue was not the `devTool` setup:
the issue was this: https://github.com/Microsoft/vscode-chrome-debug/issues/64
solution: https://stackoverflow.com/a/37903050/6126033

work client and serverside too
Sometimes clients breakpoints are not hit at the 1st visit. In that case, try to navigate away and navigate back.

**Comment by ben-siewert** [comment]
@craigpg The next.config.js recommendation is working for me when I use @AssisrMatheus launch.json config. Thanks.

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
